### PR TITLE
OADP-4935: Default permission for Secrets

### PR DESCRIPTION
**Pull request by mateusoliveira43**
## Why the changes were made

have more restrictive permissions for Pods' `volume.secrets`.

## How to test the changes made

Run `make deploy-olm` and check if nothing was broken by the change.

To test `volume.secrets`, create a DPA WITHOUT specifying credentials in BSL/VSL. Reference https://redhat-internal.slack.com/archives/C039LRSDC8Z/p1728048939142289

In master, you should have this result
```console
sh-5.1$ ls -1la credentials/cloud
lrwxrwxrwx. 1 root 1000640000 12 Oct  4 17:27 credentials/cloud -> ..data/cloud
sh-5.1$ ls -1laL credentials/cloud
-rw-r--r--. 1 root 1000640000 132 Oct  4 17:27 credentials/cloud
```
In this PR branch, you should have this result
```console
sh-5.1$ ls -1la credentials/cloud 
lrwxrwxrwx. 1 root 1000660000 12 Oct  4 13:53 credentials/cloud -> ..data/cloud
sh-5.1$ ls -1laL credentials/cloud 
-r--r-----. 1 root 1000660000 132 Oct  4 13:53 credentials/cloud
```
You also need to test operations, like backup/restores, with these changes, since E2E tests do not create DPAs  WITHOUT specifying credentials in BSL/VSL.

### Test 1: No BSL credential and no VSL

DPA
```yaml
spec:
  backupLocations:
    - velero:
        config:
          profile: default
          region: us-east-1
        default: true
        objectStorage:
          bucket: my-bucket-name
          prefix: velero
        provider: aws
  configuration:
    velero:
      defaultPlugins:
        - openshift
        - aws
```

```console
sh-5.1$ ls -1a credentials/cloud
credentials/cloud
sh-5.1$ ls -1alL credentials/cloud
-r--r-----. 1 root 1000660000 132 Oct  7 18:22 credentials/cloud
sh-5.1$ ls -1a tmp/credentials/test-oadp-operator/
.
..
sh-5.1$ ls -1alL tmp/credentials/test-oadp-operator/
total 12
drwxr-xr-x. 2 1000660000 root 4096 Oct  7 18:22 .
drwxr-xr-x. 3 1000660000 root 4096 Oct  7 18:22 ..
```

### Test 2: BSL credential and no VSL

DPA
```yaml
spec:
  backupLocations:
    - velero:
        config:
          profile: default
          region: us-east-1
        credential:
          key: cloud
          name: cloud-credentials
        default: true
        objectStorage:
          bucket: my-bucket-name
          prefix: velero
        provider: aws
  configuration:
    velero:
      defaultPlugins:
        - openshift
        - aws
```

```console
sh-5.1$ ls -1a credentials/cloud
ls: cannot access 'credentials/cloud': No such file or directory
sh-5.1$ ls -1alL credentials/cloud
ls: cannot access 'credentials/cloud': No such file or directory
sh-5.1$ ls -1a tmp/credentials/test-oadp-operator/
.
..
cloud-credentials-cloud
sh-5.1$ ls -1alL tmp/credentials/test-oadp-operator/
total 16
drwxr-xr-x. 2 1000660000 root 4096 Oct  7 18:32 .
drwxr-xr-x. 3 1000660000 root 4096 Oct  7 18:32 ..
-rw-r--r--. 1 1000660000 root  132 Oct  7 18:33 cloud-credentials-cloud
```

### Test 3: BSL credential and no VSL credential

DPA
```yaml
spec:
  backupLocations:
    - velero:
        config:
          profile: default
          region: us-east-1
        credential:
          key: cloud
          name: cloud-credentials
        default: true
        objectStorage:
          bucket: my-bucket-name
          prefix: velero
        provider: aws
  configuration:
    velero:
      defaultPlugins:
        - openshift
        - aws
  snapshotLocations:
    - velero:
        config:
          profile: default
          region: us-west-2
        provider: aws
```

```console
sh-5.1$ ls -1a credentials/cloud
credentials/cloud
sh-5.1$ ls -1alL credentials/cloud
-r--r-----. 1 root 1000660000 132 Oct  7 18:35 credentials/cloud
sh-5.1$ ls -1a tmp/credentials/test-oadp-operator/
.
..
cloud-credentials-cloud
sh-5.1$ ls -1alL tmp/credentials/test-oadp-operator/
total 16
drwxr-xr-x. 2 1000660000 root 4096 Oct  7 18:35 .
drwxr-xr-x. 3 1000660000 root 4096 Oct  7 18:35 ..
-rw-r--r--. 1 1000660000 root  132 Oct  7 18:36 cloud-credentials-cloud
```

### CODE REVIEW

File: controllers/common.go
Comment (mateusoliveira43) [inline controllers/common.go:58]: from what I searched, this is used if using CloudStorage

File: controllers/common.go
Comment (mateusoliveira43) [inline controllers/common.go:60]: these `else if`s seem strange to me. Should not they be only `if`s?

Comment (kaovilai) [reply controllers/common.go:60]: A given volume would only have one type. Using else if would probably save an unnecessary second if check.

Comment (mateusoliveira43) [reply controllers/common.go:60]: right, missed that
```go
// Represents the source of a volume to mount.
// Only one of its members may be specified.
type VolumeSource struct {
```

**Review by kaovilai — APPROVED** [review]

File: pkg/common/common.go
Comment (shubham-pampattiwar) [inline pkg/common/common.go:104]: Thinking about this, `600` means the owner can read and write the secret. Generally secrets are ment to be read-only, typically we do not modify secrets. I am thinking we can be more restrictive here, What about 400 or 440 ? just make the secret read-only.

File: controllers/common.go
Comment (shubham-pampattiwar) [reply controllers/common.go:58]: Yes we need to test this using the CloudStorage API for bucket creation and then use it in DPA BSL configuration. https://docs.openshift.com/container-platform/4.15/backup_and_restore/application_backup_and_restore/aws-sts/oadp-aws-sts.html

File: pkg/common/common.go
Comment (mateusoliveira43) [reply pkg/common/common.go:104]: I think is ok

This is applied only to Secret data inside Pod, right? User could still see Secret contents looking Secret object in cluster

**Comment by kaovilai** [comment]
re: issue around using secret for something other than object storage access key
https://github.com/vmware-tanzu/velero/issues/7767

but it's not a current concern, something if impacted can be addressed later.

**Comment by weshayutin** [comment]
weird..  I see no change to the cred's permssions. Perhaps I'm doing something wrong?
* make deploy-olm w/ * 46c1b46 - (HEAD -> fix/default-permission) fixup! fix: Default permission for Secrets (9 hours ago) <Mateus Oliveira>
* create dpa w/ creds of course
* rsh into velero pod

```
-5.1$ cd /tmp/credentials/openshift-adp/
sh-5.1$ ls -la
total 4
drwxr-xr-x. 2 1000720000 root  37 Oct  3 20:05 .
drwxr-xr-x. 3 1000720000 root  27 Oct  3 20:05 ..
-rw-r--r--. 1 1000720000 root 116 Oct  3 20:08 cloud-credentials-cloud
sh-5.1$ 

```

**Comment by weshayutin** [comment]
@shubham-pampattiwar look at the bug https://issues.redhat.com/browse/OADP-4935
The bug does not reference, what type of storage location ( bsl, vsl ), it's referring only to mounted secrets.
AFAICT this patch has not fixed the jira issue yet.

**Comment by mateusoliveira43** [comment]
@weshayutin @shubham-pampattiwar for this to work, you can not set `dpa.spec.backupLocation.credentials`

In my test (but it is wrong still, it is `0440` and it is set as `0400` in code)
```console
sh-5.1$ ls -1la credentials/cloud 
lrwxrwxrwx. 1 root 1000660000 12 Oct  4 13:53 credentials/cloud -> ..data/cloud
sh-5.1$ ls -1laL credentials/cloud 
-r--r-----. 1 root 1000660000 132 Oct  4 13:53 credentials/cloud
sh-5.1$ ls -1a tmp/credentials/test-oadp-operator/
.
..
```

**Comment by sseago** [comment]
@mateusoliveira43 Hmm. For that last example, that's root-owned -- so the velero pod really needs the group read to be able to access it since it doesn't run as root.

**Review by sseago — DISMISSED** [review]

**Review by kaovilai — DISMISSED** [review]

**Comment by mateusoliveira43** [comment]
/hold
lets test more just to be sure, before merging?

**Comment by kaovilai** [comment]
it won't merge until all required tests green

**Comment by kaovilai** [comment]
the lgtm will just enable auto retests

**Comment by mateusoliveira43** [comment]
my worries here is that E2E does not test these changes, as I added in PR description

**Comment by weshayutin** [comment]
@kaovilai @sseago so when I pull this and get on the cluster even w/ my standard dpa that I used previously .. I will see the permissions changed?

**Comment by mateusoliveira43** [comment]
@weshayutin if you specify credentials in BSL (even if the name is the default one), no

added info in PR description

**Comment by weshayutin** [comment]
@mateusoliveira43 In your opinion does that fix the jira issue?

**Comment by weshayutin** [comment]
ok.. lgtm to w/ my default dpa, same one I tested before
```
sh-5.1$ 
sh-5.1$ ls -la /credentials/cloud 
lrwxrwxrwx. 1 root 1000720000 12 Oct  4 19:55 /credentials/cloud -> ..data/cloud
sh-5.1$ ls -la /credentials/..data/cloud 
-r--r-----. 1 root 1000720000 116 Oct  4 19:55 /credentials/..data/cloud
sh-5.1$ 
```
```
spec:
  backupLocations:
  - velero:
      config:
        profile: default
        region: us-west-2
      credential:
        key: cloud
        name: cloud-credentials
      default: true
      objectStorage:
        bucket: cvpbucketuswest2
        prefix: velero
      provider: aws
  configuration:
    nodeAgent:
      enable: true
      uploaderType: kopia
    velero:
      defaultPlugins:
      - kubevirt
      - csi
      - openshift
      - aws
  snapshotLocations:
  - velero:
      config:
        profile: default
        region: us-west-2
      provider: aws
```

**Comment by mateusoliveira43** [comment]
Tested BSL (WITHOUT credentials) by running modified E2E tests locally, tests passed. Did not test VSL

**Comment by shubham-pampattiwar** [comment]
### Test 1
```
spampatt@spampatt-mac oadp-operator [secret-read-only]$ oc exec -it velero-7758ddd869-fnqn7 -- /bin/sh
Defaulted container "velero" out of: velero, openshift-velero-plugin (init), velero-plugin-for-aws (init)
sh-5.1$ ls -la credentials/cloud 
lrwxrwxrwx. 1 root root 12 Oct  8 00:17 credentials/cloud -> ..data/cloud
sh-5.1$ ls -lalL credentials/cloud 
-r--r-----. 1 root root 116 Oct  8 00:17 credentials/cloud
sh-5.1$ ls -lalL tmp/credentials/openshift-adp/
total 0
drwxr-xr-x. 2 nobody nobody  6 Oct  8 00:17 .
drwxr-xr-x. 3 nobody nobody 27 Oct  8 00:17 ..
```

### Test 2
```
spampatt@spampatt-mac oadp-operator [secret-read-only]$ oc exec -it velero-5689985869-vn6b6 -- /bin/sh
Defaulted container "velero" out of: velero, openshift-velero-plugin (init), velero-plugin-for-aws (init)
sh-5.1$ ls -la credentials/cloud
ls: cannot access 'credentials/cloud': No such file or directory
sh-5.1$ ls -la tmp/credentials/openshift-adp/
total 4
drwxr-xr-x. 2 nobody nobody  37 Oct  8 00:21 .
drwxr-xr-x. 3 nobody nobody  27 Oct  8 00:21 ..
-rw-r--r--. 1 nobody nobody 116 Oct  8 00:21 cloud-credentials-cloud
sh-5.1$ 

```

### Test 3
```
spampatt@spampatt-mac oadp-operator [secret-read-only]$ oc exec -it velero-7758ddd869-5fr8m -- /bin/sh
Defaulted container "velero" out of: velero, openshift-velero-plugin (init), velero-plugin-for-aws (init)
sh-5.1$ ls -la credentials/
..2024_10_08_00_27_11.1709205179/ ..data/                           cloud                             
sh-5.1$ ls -la credentials/cloud 
lrwxrwxrwx. 1 root root 12 Oct  8 00:27 credentials/cloud -> ..data/cloud
sh-5.1$ ls -lalL credentials/cloud 
-r--r-----. 1 root root 116 Oct  8 00:27 credentials/cloud
sh-5.1$ ls -lalL tmp/credentials/openshift-adp/
total 4
drwxr-xr-x. 2 nobody nobody  37 Oct  8 00:27 .
drwxr-xr-x. 3 nobody nobody  27 Oct  8 00:27 ..
-rw-r--r--. 1 nobody nobody 116 Oct  8 00:29 cloud-credentials-cloud
```

Matches the results in PR description for the 3 test scenarios.

File: controllers/common.go
Comment (mateusoliveira43) [reply controllers/common.go]: removed change in Projected volume

**Comment by mateusoliveira43** [comment]
/unhold
discussed in scrum

**Review by kaovilai — APPROVED** [review]

**Review by sseago — APPROVED** [review]

**Review by shubham-pampattiwar — APPROVED** [review]

**Pull request merged**
